### ISSUE TITLE
Feature request: validate kresd extraConfig syntax at build time

### ISSUE THREAD

**Issue by symphorien** [NixOS/nixpkgs#124155]
## Issue description

I'm unfamiliar with lua and mistakenly used `#` to introduced comments, rebuilt and lost dns access. It's a bit annoying. Maybe the syntax could be checked for obvious mistakes at build time instead?

### Steps to reproduce

```nix
{
services.kresd = {
enable = true;
extraConfig = "# this is not a comment";
};
}
```

## Technical details

cc @vcunat as maintainer

**Comment by vcunat** [comment]
Well, no... the problem is that kresd configuration is completely general lua script; it can do basically anything and lua has very permissive syntax (typos in variables are often not an error).  I guess luacheck would help (after setting up) but only a little.

The flexibility sounded nice originally, but practice has shown us that for most people the price is too high, so we (upstream) are in the process of designing a different configuration format.  The plans are to use standard YAML with additional checks, so there it will be much easier and I'll certainly want it for the NixOS service.  (Extra lua without checks should still be possible in some way, if people need something very specific.)

**Comment by vcunat** [comment]
I also expect that with the new format, approach like `lib.toYAML { /*all config in nix attrsets*/ }` will be more comfortable to many Nixers.

**Comment by symphorien** [comment]
The thing is I am currently using coroutines to load a very big bocklist with policy.rpz off the main thread (not sure if it's the right term) so I really need lua.


> typos in variables are often not an error)

well in that case some amount of checks would be even more welcome...

**Comment by vcunat** [comment]
Oh, that's quite advanced.  Better support for big policies is also being prepared, but both projects are really big and thus uncertain when they'll land.

Perhaps I can come with some experimental `luacheck` config soon.  We're already using something in CI on the inner lua files (e.g. modules).

**Comment by mweinelt** [comment]
> The flexibility sounded nice originally, but practice has shown us that for most people the price is too high, so we (upstream) are in the process of designing a different configuration format. The plans are to use standard YAML with additional checks, so there it will be much easier and I'll certainly want it for the NixOS service. (Extra lua without checks should still be possible in some way, if people need something very specific.)

Is that something that could be considered for knot as well?

**Comment by vcunat** [comment]
There the configuration file is simple (restricted YAML) and we already have the checks: https://github.com/NixOS/nixpkgs/blob/release-20.09/nixos/modules/services/networking/knot.nix#L13

**Comment by mweinelt** [comment]
I'd rather call it extended YAML, given that you can give the same dict key multiple times and the order of keys matters AFAIK. This is the part I'm a bit saddened by.

**Comment by vcunat** [comment]
Perhaps you'd be interested in this prototype?  https://github.com/NixOS/nixpkgs/pull/81460  I wrote it after some complaints around these issues when generating knotd config from nix, but then there was no feedback...

**Comment by mweinelt** [comment]
Yes, we touched on that topic a while ago on over gitter. https://gitter.im/CZ-NIC/knot?at=5fac8d16d37a1a13d69f0166

**Comment by vcunat** [comment]
Ah, right, my memory wasn't sure of it.
